applications/smart-distancing/face_mask_recognition/utils.py

```diff
@@ -0,0 +1,140 @@
+from keras import backend as K
+from tensorflow.python.saved_model import builder as saved_model_builder, tag_constants
+from tensorflow.python.saved_model.signature_def_utils import predict_signature_def
+from sklearn.metrics import confusion_matrix
+import itertools
+import matplotlib.pyplot as plt
+from keras.preprocessing.image import ImageDataGenerator
+import tensorflow as tf
+import numpy as np
+import os
+import sys
+from keras.models import load_model
+
+FLAGS = tf.flags.FLAGS
+
+
+def plot_confusion_matrix(cls_true, cls_pred):
+    """
+    Plot confusion matrix
+    Args:
+        cls_true: Array of labels, each element in the array indicates a class id
+        cls_pred: Array of prediction results, each element in the array indicates a predicted class id
+
+    Returns:
+
+    """
+    cm = confusion_matrix(y_true=cls_true,
+                          y_pred=cls_pred,
+                          )
+    plt.ion()
+    plt.matshow(cm, cmap=plt.cm.Reds)
+    # Make various adjustments to the plot.
+    plt.colorbar()
+    tick_marks = np.arange(FLAGS.no_classes)
+    plt.xticks(tick_marks, FLAGS.classes)
+    plt.yticks(tick_marks, FLAGS.classes)
+    plt.xlabel('Predicted')
+    plt.ylabel('True')
+
+    for i, j in itertools.product(range(cm.shape[0]), range(cm.shape[1])):
+        plt.text(j, i, "{}".format(cm[i, j]),
+                 horizontalalignment="center",
+                 color="black")
+        # Ensure the plot is shown correctly with multiple plots
+        # in a single Notebook cell.
+    plt.show()
+    plt.savefig(os.path.join(FLAGS.result_dir, 'confusion_matrix.png'))
+
+
+def data_reader(path, batch_size, input_size):
+    """
+    Using keras image generator to read images
+    Args:
+        path: Image path (each folder at this path will be considered as a class)
+        batch_size: Batch size
+        input_size: The size of the network input
+
+    Returns:
+        data: A generator of images
+
+    """
+    data_generator = ImageDataGenerator(rescale=1.0 / 255)
```

Comment:
using `tf.data` API is much more faster than `ImageDataGenerator` since `ImageDataGenerator` is implemented in python and Numpy but `tf.data` has C runtime and export the computations to a computational  graph.

Reply:
Yes, but the training process for this task is not bottleneck because of the size of the network. I will take 2-5 minutes to be trained.
ImageDataGenerator is easy-to-use for this task, and the issue of speed is solved at tensorflow nightly version. We can upgrade this issue at new version of TF.
By the way, I try to solve it with tf.data